### PR TITLE
Add new tool to check HTML

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Re-opening of https://github.com/rust-lang/rust/pull/84480.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I'd like this to use the same logic as added by https://github.com/rust-lang/rust/pull/86152 (once it lands) to make sure we're lazily evaluating the condition here. Otherwise broadly looks good but I'd like to take another look before final r+ once this is updated.

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -161,6 +161,54 @@ You can skip linkcheck with --exclude src/tools/linkchecker"
...
+        }
+        // Ensure that a few different kinds of documentation are available.
+        builder.default_doc(&[]);
+        builder.ensure(crate::doc::Rustc { target: self.target, stage: builder.top_stage });
+
+        try_run(
+            builder,
+            builder
+                .tool_cmd(Tool::HtmlChecker)
+                .arg(builder.out.join(self.target.triple).join("doc")),
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:207]: `builder.doc_out(self.target)` should be more appropriate here.

**Review by Mark-Simulacrum — APPROVED** [review]
r=me with nit fixed

File: src/bootstrap/test.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs:207]: :+1:

**Comment by Mark-Simulacrum** [comment]
If you could squash commits that would also be good.

**Comment by GuillaumeGomez** [comment]
Doing that!

@bors: r-

**Comment by GuillaumeGomez** [comment]
I kept the "specialized" commits on their own, merged the others together.

@bors: r=Mark-Simulacrum

**Pull request merged**